cli/src/commands/parse/msgs.rs

```diff
@@ -0,0 +1,426 @@
+    })
+}
+
+fn remove_content_headers(headers_string: String) -> Result<String> {
+    let mut clean_headers_string: String;
+
+    clean_headers_string = CONTENT_TYPE_MIME_HEADER_RX
+        .clone()
+        .replace(&headers_string, "")
+        .to_string();
+
+    clean_headers_string = CONTENT_TRANSFER_ENCODING_MIME_HEADER_RX
+        .clone()
+        .replace(&clean_headers_string, "")
+        .to_string();
+
+    Ok(clean_headers_string)
+}
+
+fn read_msg_to_document(path: &PathBuf) -> Result<Document> {
+    if !path.is_file() {
+        return Err(anyhow!("No such file: {:?}", path));
+    }
+
+    let mut compound_file = cfb::open(path)?;
+
+    // Headers
+    let headers_string =
+        read_unicode_stream_to_string(STREAM_PATH_MESSAGE_HEADER.clone(), &mut compound_file)?;
+
+    // As the content type won't match the parsed value from the body in the msg
+    let headers_string_no_content_headers = remove_content_headers(headers_string)?;
+
+    let plain_body_string =
+        read_unicode_stream_to_string(STREAM_PATH_MESSAGE_BODY_PLAIN.clone(), &mut compound_file)?;
+
+    // Attachments
+    let mut attachment_number = 0;
+    let mut attachments = Vec::new();
+    loop {
+        let attachment_path = get_attachment_store_path(attachment_number);
+
+        if compound_file.is_storage(attachment_path.clone()) {
+            attachments.push(read_attachment(attachment_path, &mut compound_file)?);
+        } else {
+            break;
+        }
+
+        attachment_number += 1;
+    }
+
+    // User Properties
+    let mut user_properties = PropertyMap::new();
+    user_properties.insert_string(
+        MSG_NAME_USER_PROPERTY_NAME.to_string(),
+        path.file_name()
+            .context("Could not get file name")?
+            .to_string_lossy()
+            .to_string(),
+    );
+
+    Ok(Document {
+        raw_email: RawEmail {
+            body: RawEmailBody::Plain(plain_body_string),
+            headers: RawEmailHeaders::Raw(headers_string_no_content_headers),
+            attachments,
+        },
+        user_properties,
+        comment_id: None,
+    })
+}
+
+fn upload_batch_of_documents(
+    client: &Client,
+    source: &Source,
+    documents: &Vec<Document>,
+    transform_tag: &TransformTag,
+    no_charge: bool,
+    statistics: &Arc<Statistics>,
+) -> Result<()> {
+    client.sync_raw_emails(
+        &source.full_name(),
+        documents,
+        transform_tag,
+        false,
+        no_charge,
+    )?;
+    statistics.add_uploaded(documents.len());
+    Ok(())
+}
+
+pub fn get_msgs_in_directory(directory: &PathBuf) -> Result<Vec<DirEntry>> {
+    Ok(std::fs::read_dir(directory)?
+        .filter_map(|path| {
+            if let Ok(path) = path {
+                if !path.path().extension().is_some_and(|msg| msg == "msg") {
+                    None
+                } else {
+                    Some(path)
+                }
+            } else {
+                None
+            }
+        })
```

Comment:
Written slightly strangely - nested `if` normally indicates you can use a combinator, in this case `.ok()`. Also you negate the if statement for some reason, which makes it harder to read?

```suggestion
        .filter_map(|path| {
            let path = path.ok()?;
            if path.path().extension().is_some_and(|msg| msg == "msg") {
                Some(path)
            } else {
                None
            }
        })
```